[FAB-17906] Implement snapshotRequestBookKeeper

wenjianqiao opened the pull request:
Signed-off-by: Wenjian Qiao <wenjianq@gmail.com>

#### Type of change
- New feature

#### Description
Add a bookkeeper db to store snapshot requests and add/delete/list
snapshot requests. It also maintains smallest height for all the
pending snapshot requests.

#### Additional details
This is part of FAB-17906. In a separate PR, the snapshot mgmt APIs will be implemented and call the snapshot request bookkeeper to submit/cancel/list snapshot requests.

#### Related issues
https://jira.hyperledger.org/browse/FAB-17906

wenjianqiao committed:
[FAB-17906] Implement snapshotRequestBookKeeper

Add a bookkeeper db to store snapshot requests and add/delete/list
snapshot requests. It also maintains smallest height for all the
pending snapshot requests.

Signed-off-by: Wenjian Qiao <wenjianq@gmail.com>

manish-sethi commented on core/ledger/kvledger/snapshot_mgmt.go:
```diff
@@ -60,3 +62,130 @@ func (l *kvLedger) ListSnapshots() ([]string, error) {
+
+	smallestRequestHeight := defaultSmallestHeight
+	if itr.Next() {
+		if err := itr.Error(); err != nil {
```
could cause miss of an error

manish-sethi commented on core/ledger/kvledger/snapshot_mgmt.go:
```diff
@@ -60,3 +62,130 @@ func (l *kvLedger) ListSnapshots() ([]string, error) {
+}
+
+// add adds the given height to the bookkeeper db and returns an error if the height already exists
+func (bk *snapshotRequestBookkeeper) add(height uint64) error {
```
Nit picking: Idiomatically, you would use one letter name for the function receiver. You may like to change `bk` to `b` or `k`

manish-sethi commented on core/ledger/kvledger/snapshot_mgmt_test.go:
```diff
@@ -0,0 +1,128 @@
+	bcInfo, _ := l.GetBlockchainInfo()
+	require.Equal(t, &common.BlockchainInfo{
+		Height: 1, CurrentBlockHash: gbHash, PreviousBlockHash: nil,
+	}, bcInfo)
```
Though this code is not related and you could directly create an instance. However, since you have it here, it will be good to add a case for error propagation during ledger opening.

manish-sethi commented on core/ledger/kvledger/snapshot_mgmt_test.go:
```diff
@@ -0,0 +1,128 @@
+	require.ElementsMatch(t, requestHeights, []uint64{})
+}
+
+func TestSnapshotRequestBookKeeper_ErrorPaths(t *testing.T) {
```
Nit: `_` in function name is not idiomatic

wenjianqiao commented:
@manish-sethi Thank you. I have addressed your comments and also simplified unit tests by directly creating a snapshotRequestBookkeeper. The unit test to kvledger change will be covered when snapshot mgmt APIs are implemented.
